### DISCUSSION TITLE
Trouble Installing Homebrew in MacOSX 10.11.6

### DISCUSSION THREAD

**Discussion by Mishapen-Sprocket** [Homebrew/discussions#1141] [Getting started]
EDIT: I'm back! Reinstalling my OS did not fix my PAM issue. :(

But I ran 
```
mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
```

not knowing what it would do and I got hombrew in a nice folder in my user folder. However, when I try to run a brew command I get 

```
-bash: brew: command not found
```

So where do I put the hombrew folder so my computer has its commands? You know, in a way that doesn't need a working sudo command?

(I know the best place to put it is /usr/local but *see the rest of this discussion* for why I can't do that.)

------- original post --------------

I have a Macbook running "El Capitan" 10.11.6 . The processor is a 2.4 GHz Intel Core 2 Duo

I opened up terminal and gave it the command:

/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"

Like the instructions said to, and got the following error message:

==> Checking for `sudo` access (which may request your password).
sudo: unable to initialize PAM: No such file or directory
Need sudo access on macOS (e.g. the user [my user name here] needs to be an Administrator)!

I thought the problem was that my version of OSX doesn't have the directory usr/local to install in, so I created a directory called local and put in a folder called usr I made in ~ but that didn't change the error message.

I've checked in system preferences and I am an admin on this account. 

So what's the issue here?

**Comment by Mishapen-Sprocket** [comment]
I made a typo: the path for local is ~/usr/local, even though I wrote it backwards in the post. 

I also have never installed homebrew on this computer before, so that's not an issue. I tried running the command "brew" and got "-bash: brew: command not found" which if I understand it right means homebrew isn't already installed.

**Comment by carlocab** [comment]
I'm not familiar with the details, but my (possibly mistaken) understanding is that the installer uses features of `sudo` that doesn't quite work on older versions of macOS. What's the output of the following commands?

```
file /etc/pam.d/sudo
cat /etc/pam.d/sudo
```

If you'd rather not work out these `sudo` issues, you may wish to try using an alternative installation method instead: https://docs.brew.sh/Installation#untar-anywhere

I should also mention that the earliest supported macOS version is Mojave. This means that most, if not all, of the software you install using Homebrew will require building them from source, which often isn't a great experience. (It's not ideal, but our resources are too limited to support more than the three most recent versions of macOS.) You may wish to consider an alternative package manager (e.g. MacPorts) that offers support for macOS versions before Mojave.

**Reply by Mishapen-Sprocket** [reply]
file /etc/pam.d/sudo 

yields

/etc/pam.d/sudo: ASCII text

cat /etc/pam.d/sudo

yields

# sudo: auth account password session
auth       sufficient     pam_securityserver.so
auth       sufficient     pam_unix.so
auth       required       pam_deny.so
account    required       pam_permit.so
password   required       pam_deny.so
session    required       pam_permit.so

**Reply by carlocab** [reply]
Ok, what about

```
ls -l /usr
ls -l /usr/local
```
?

By the way, you can put three backticks (\`\`\`) on the line before and the line after you paste output to format them properly.

**Reply by Mishapen-Sprocket** [reply]
Thanks for the tip. I checked the package I want to install and my OS isn't supported anyway so it looks like I'm going to have to figure out how to build it from source (which seems beyond my capabilities) or find an alternative program, so I'm going to do the latter!

**Reply by Mishapen-Sprocket** [reply]
``` 
ls –l /usr
```
yields

```
total 8
drwxr-xr-x     5 root  wheel    170 Jul 31  2015 X11
lrwxr-xr-x     1 root  wheel      3 Mar 17  2016 X11R6 -> X11
drwxr-xr-x     3 root  wheel    102 Dec 10  2015 adic
drwxr-xr-x  1056 root  wheel  35904 Jul 18  2018 bin
drwxr-xr-x   266 root  wheel   9044 Jul 18  2018 lib
drwxr-xr-x   186 root  wheel   6324 Jul 18  2018 libexec
drwxr-xr-x     3 root  wheel    102 Jul 20  2017 local
drwxr-xr-x   243 root  wheel   8262 Jul 18  2018 sbin
drwxr-xr-x    45 root  wheel   1530 Mar 17  2016 share
drwxr-xr-x     4 root  wheel    136 Jan 13  2016 standalone

```

```
ls -l usr/local

```

yields nothing (no error message, just starts a new command line)

**Reply by carlocab** [reply]
You can try to do
```
sudo chown $(whoami) /usr/local
```
and, if that succeeds, the installer might work.

**Reply by Mishapen-Sprocket** [reply]
Tried that. Got

```
sudo: unable to initialize PAM: No such file or directory

```

**Reply by carlocab** [reply]
What about `ls -l /etc/pam.d`?

**Reply by Mishapen-Sprocket** [reply]
total 96
-rw-r--r--  1 root  wheel  264 Nov 17  2015 authorization
-rw-r--r--  1 root  wheel  181 Nov 17  2015 checkpw
-r--r--r--  1 root  wheel  102 Nov 29  2007 chkpasswd
-rw-r--r--  1 root  wheel  280 Jan 13  2008 cups
-rw-r--r--  1 root  wheel  243 Nov 25  2007 ftpd
-r--r--r--  1 root  wheel  362 Nov 29  2007 login
-r--r--r--  1 root  wheel  408 Jun 17  2009 login.serverinstall
-rw-r--r--  1 root  wheel  152 Dec 10  2015 login.term
-r--r--r--  1 root  wheel  191 Nov 29  2007 other
-r--r--r--  1 root  wheel  185 Nov 29  2007 passwd
-r--r--r--  1 root  wheel  211 Jul 31  2015 rshd
-r--r--r--  1 root  wheel  141 Nov 29  2007 samba
-rw-r--r--  1 root  wheel  408 Oct 24  2015 screensaver
-rw-r--r--  1 root  wheel  140 Jul 31  2015 smbd
-r--r--r--  1 root  wheel  403 Nov 29  2007 sshd
-r--r--r--  1 root  wheel  449 Nov 29  2007 sshd.serverinstall
-r--r--r--  1 root  wheel  343 Nov 29  2007 su
-r--r--r--  1 root  wheel  280 Nov 29  2007 sudo

**Reply by davidprush** [reply]
Are you logged in with a user that had root privileges? 

Possible solutions:

https://apple.stackexchange.com/a/190884

https://apple.stackexchange.com/a/388962

Try those and report back here.

**Reply by Mishapen-Sprocket** [reply]
It occurs to me the following details might help:
This is a used computer. 
When I got it, the first thing it did was have me set up an account (called "me"). 
Then I imported all the files from my old computer (also OSX) into a new account (called "old me"). 
Finally, I gave admin privileges to "old me" using System Preferences
I've been logged into "old me" ever since. 

Should I switch accounts back to "me"? Does "old me" not have root privileges? If I install a package while logged into in "me" can I use it from terminal in "old me"?

I'm reading the links now.

**Reply by carlocab** [reply]
It's possible. Try doing the following. While logged into "old me" enter
```
whoami
```
then remember the output of that. Then, log into "me", and do
```
sudo chown <output of whoami from above> /usr/local
```
If/when it asks for a password, be sure to use the password for "me" and not "old me". Then, log back into "old me" and try to install Homebrew again.

I'm not certain this will work, because the error message reads to me as more than just an insufficient permissions problem (which my suggestion above would get around). However, I guess it might be worth a try.

**Reply by Mishapen-Sprocket** [reply]
I read the links. The id command, whoami, systempreferences advanced options all agree my user name is [redacted], which is the same username that's been showing up in my terminal this whole time:
```
[redacted name of my computer]: ~ [redacted username]$
```

```
$ sudo visudo /private/etc/sudoers
```

and

```
$ sudo visudo /etc/sudoers
```

both yield 

```
sudo: unable to initialize PAM: No such file or directory
```
I logged into "me," did 

```

sudo chown [redacted] /usr/local
```

and got

```
sudo: unable to initialize PAM: No such file or directory
```

**Reply by carlocab** [reply]
Yea, your `sudo` command is broken, but it could be for any number of reasons. The simplest solution might be a fresh install of the newest macOS compatible with your Mac.

Failing that, I'd start by Googling the error message `sudo: unable to initialize PAM: No such file or directory` and see where that leads. 

FYI, I don't really recommend the solution in [this StackExchange answer](https://unix.stackexchange.com/a/77425) which shows up for me as a search result to the query mentioned above. However, if you're desperate and careful enough, it might work for you.

**Reply by BlackLionFounder** [reply]
> I should also mention that the earliest supported macOS version is Mojave. This means that most, if not all, of the software you install using Homebrew will require building them from source, which often isn't a great experience. (It's not ideal, but our resources are too limited to support more than the three most recent versions of macOS.) You may wish to consider an alternative package manager (e.g. MacPorts) that offers support for macOS versions before Mojave.

Thank you for stating this, I could not find any information until this post. Thank you.

**Reply by carlocab** [reply]
The macOS requirements are documented here: https://docs.brew.sh/Installation#macos-requirements

Note that my original comment is now outdated: we only build bottles (binary packages) for Catalina and newer now (since the release of Monterey). More generally, we build bottles only for the last three versions of macOS, in line with the versions that Apple still supports.

**Comment by Mishapen-Sprocket** [comment]
Alright. I'm reinstalling El Capitan to see if I can get a fresh set of PAM files.

The weird thing is before today, I never did anything involving sudo. I wonder if the files got messed up by a random decay in flash memory where they're stored. 

Some of the google results for the PAM error mention computers with accounts restored from time machine files, like mine. Maybe there's a bug in Time Machine that messes up PAM files when a backup from one computer is restored into a different one.

**Reply by carlocab** [reply]
You did say the Mac was used. `sudo` could've been broken already when you got it. You just never noticed because you hadn't used it before.

**Comment by Mishapen-Sprocket** [comment]
I'm back! Reinstalling my OS did not fix my PAM issue. :(

But I ran 
```
mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
```

not knowing what it would do and I got hombrew in a nice folder in my user folder. However, when I try to run a brew command I get 

```
-bash: brew: command not found
```

So where do I put the hombrew folder so my computer has its commands? You know, in a way that doesn't need a working sudo command?

(I know the best place to put it is /usr/local but *see the rest of this discussion* for why I can't do that.)

**Reply by carlocab** [reply]
You need to do
```sh
export PATH="/path/to/homebrew/bin:$PATH"
```
where `/path/to/homebrew` is the full path to the `homebrew` directory. You may wish to add this to your `.bash_profile` (or `.zshrc`) so that you don't need to run it every time you open the terminal

**Reply by Mishapen-Sprocket** [reply]
Thanks a million! Seems good so far, but to install the brew I want, I need git in the path. 

I have git installed (and commands run successfully), but I can't figure out where! How do I get git in the same path as homebrew?

EDIT: Oh frig, it's in usr/bin/git. I have to reinstall git in ~/ (where I have homebrew).

**Reply by Mishapen-Sprocket** [reply]
Can I export a path to git or something so it ends up in the same PATH as homebrew?

**Reply by carlocab** [reply]
`git` is in your `PATH`. The command
```sh
export PATH="/path/to/homebrew/bin:$PATH"
```
prepends `/path/to/homebrew/bin` to your existing `PATH` (which should include `git`). If it complains about not finding `git`, you may not have the Command Line Tools installed. If
```sh
which xcode-select
```
returns a path (as opposed to `xcode-select not found`), then you can just do
```sh
xcode-select --install
```
If not, you'll probably need to install them from https://developer.apple.com/download/more/. (You'll need to sign up for an Apple developer account.) You're looking for the Command Line Tools for Xcode 7.3.1.

**Reply by Mishapen-Sprocket** [reply]
I think I have it.

I'm not sure exactly what happened, but it looks like the "git not in path" error can be ignored. I think the actual problem is that when I tried to "brew install" the package I want in homebrew, one of the 3 dependencies it tried to download failed to download properly. The second attempt to "brew install" worked. 

I have xcode 8.0, which is not up to date, and it's causing a warning, but it doesn't look like I can update it on this OS and it seems like git is working anyway. 

I've having problems using the package I brew installed, but it *does* have a folder in homebrew/cellar. I think I have homebrew up and running now and my new problem is one level deeper (not a problem for this forum).

To sum up:

Running
```
/bin/bash -c "$(curl -fsSL https://raw.githubusercontent.com/Homebrew/install/HEAD/install.sh)"
```

failed because my sudo command is broken because of a messed-up PAM file

So I needed to install outside usr/local using

```
mkdir homebrew && curl -L https://github.com/Homebrew/brew/tarball/master | tar xz --strip 1 -C homebrew
```
and then
```
export PATH="~/homebrew/bin:$PATH"

```
(homebrew was in my user directory ~ ).

**Reply by carlocab** [reply]
Great; glad you're up and running. As I imagine you've already noticed, most things you `brew install` will need to build from source, and this requires some patience (and inevitable difficult-to-diagnose build failures).

However, there are still ~900 formulae bottled for El Capitan, and you should be able to install those without having to do a source build. You can see which formulae those are with
```sh
cd $(brew --repo homebrew/core)/Formula
grep -rl 'el_capitan:' .
```

**Reply by gromgit** [reply]
> export PATH="~/homebrew/bin:$PATH"

I'd actually recommend doing this instead:
```
eval "$(~/homebrew/bin/brew shellenv)"
```
`brew shellenv` issues the correct commands to set up your shell environment for Homebrew, including `MANPATH` and `INFOPATH` (so you can access whatever docs are installed with the formulae).

**Reply by carlocab** [reply]
I don't know when the last time Apple updated `man` on macOS, but the version on Catalina doesn't need `MANPATH` to be set. It uses your `PATH` to find a matching `share/man` directory on its own. This might have been true as early as El Capitan.
